### PR TITLE
V4

### PR DESCRIPTION
**Pull request by EmptyInfinity**

### CODE REVIEW

**Review by if0s — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -32,29 +33,40 @@
...
-    "typescript": "3.8.3"
+    "nyc": "15.1.0",
+    "sinon": "11.1.2",
+    "ts-node": "10.2.1",
+    "typescript": "4.4.3"
   },
   "dependencies": {
     "@elastic.io/bunyan-logger": "1.0.8",
     "@types/sinon": "10.0.0",
+    "aws-sdk": "2.1152.0",
```
[the remark this answers is not in the document]
Comment (EmptyInfinity) [reply package.json:65]: oops

**Review by if0s — APPROVED** [review]

**Review by if0s — APPROVED** [review]

**Comment by EmptyInfinity** [comment]
> **Platform team requirements:**
> 
- [x] 1. Nightly tests in CI.
- [x] 2. Disable dev dependencies check during audit.
- [x] 3. Version should adhere to semver rules. Breaking change - new major version.
- [x] 4. All tests from `object-storage-client` must be present here. Especially which cover some edge cases (like passing the same stream on retry). 
(Passing same stream could be covered only with [positive test](https://github.com/elasticio/maester-client/blob/v4/spec/StorageClient.spec.ts#L64))
- [x] 5. Exponential back-off like in `object-storage-client`.
(Because of having ability to config "delay on retry" was implemented logic like [this](https://github.com/elasticio/maester-client/blob/v4/src/utils.ts#L42))
- [ ] 6. All errors should be handled like in latest `object-storage-client`. Now they don't. E.g. I don't see Client error handler.
(Error handling were left as it is, you may familiar with handling approach in [integration-tests](https://github.com/elasticio/maester-client/blob/v4/spec-integration/ObjectStorage.spec.ts#L183))

Now related to how our services are using the lib 
- [x] 1. We should be able to create an instance to make add/update/delete requests. The instance should accept maester uri and jwtSecret.
- [x] 2. Each method that modifies DB object should accept `ttl`. `ttl` can have `-1` value.
- [x] 3. There should be ability to add stream for encrypt/decrypt.
- [x] 4. There should be ability to GET object as stream with specified jwt payload.
- [x] 5. There should be ability to PUT as stream with retries: 0.
- [x] 6. There should be ability to PUT with custom ttl and contentType (we get it from webhook request headers).
 (currently, there is no way to pass custom contentType, because it's [calculated automatically](https://github.com/elasticio/maester-client/blob/v4/src/StorageClient.ts#L22))
- [x] 7. There should be ability to POST a json object.
- [x] 8. There should be ability to GET object as json object.
([ObjectStorage](https://github.com/elasticio/maester-client/blob/v4/src/ObjectStorage.ts#L68) will return a string. And [ObjectStorageWrapper](https://github.com/elasticio/maester-client/blob/v4/src/ObjectStorageWrapper.ts#L63) will return parsed object (this class is designed to work with JSON objects)
- [x] 9. There should be ability to DELETE an object by ID.

**Comment by PaulAnnekov** [comment]
Can you provide an example on how to override `Content-Type` for `update`?

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  return {
+    retryDelay: (retryDelay > retryDelay_MAX_LIMIT || retryDelay < retryDelay_MIN_LIMIT) ? REQUEST_RETRY_DELAY : retryDelay,
+    retriesCount: (retriesCount > retriesCount_MAX_LIMIT || retriesCount < retriesCount_MIN_LIMIT) ? REQUEST_MAX_RETRY : retriesCount,
+    requestTimeout: (requestTimeout > requestTimeout_MAX_LIMIT || requestTimeout < requestTimeout_MIN_LIMIT) ? REQUEST_TIMEOUT : requestTimeout
+  };
+};
+
+// + 1s for each retry
+export const getDelayTime = (retryDelay: number, currentReties: number): number => {
```
Comment (PaulAnnekov) [inline src/utils.ts:53]: Please, read about exponential backoff and use it instead. You can check in monorepo how we implemented it for our version of the lib.

File: src/StorageClient.ts [hunk 3]
```diff
@@ -39,97 +45,132 @@ export class StorageClient {
...
+        if (process.env.NODE_ENV === 'test') {
+          log.trace(retryDelay, retriesCount, requestTimeout);
+        }
+        res = await this.api.request({ ...axiosReqConfig, data: bodyAsStream, timeout: requestTimeout });
       } catch (e) {
+        if (e instanceof ObjectStorageClientError) {
+          throw e;
+        }
         err = e;
+        if (err?.response?.status < 500) throw e; // The request was made and the server responded with a status code
```
Comment (PaulAnnekov) [inline src/StorageClient.ts:77]: Just use "validateStatus: null", because you validate it manually, you don't need the help of axios.

File: README.md [hunk 4]
```diff
@@ -4,16 +4,21 @@ The official Elastic.io object-storage client.
 
 ## Usage
 
+This library propose you two clients: `ObjectStorage` & `ObjectStorageWrapper`.
```
Comment (PaulAnnekov) [inline README.md:7]: It's nice to have a section about how to handle errors. You can check small section in our lib for reference.

**Review by PaulAnnekov — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 5]
```diff
@@ -2,7 +2,7 @@ version: 2.1
 jobs:
   build:
     docker:
-      - image: circleci/node:16-stretch
+      - image: cimg/node:16.16.0
```
Comment (PaulAnnekov) [inline .circleci/config.yml:5]: No, 12.3.0. It should work under nodejs 12.3.0

File: src/interfaces.ts [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
 
-export enum ObjectHeaders {
-  ttl = 'x-eio-ttl'
-}
-
 export interface reqWithBodyHeaders {
-  ttl?: number;
-  override?: RequestHeaders;
+  [TTL_HEADER]?: number;
+  'content-type'?: string;
```
Comment (PaulAnnekov) [inline src/interfaces.ts:13]: it will be nice to see it as exported const too, as TTL

**Review by PaulAnnekov — CHANGES_REQUESTED** [review]
